test/lock_free_queue_tests.cpp

```diff
@@ -0,0 +1,441 @@
+  ASSERT_TRUE(buffer.empty()) << "Buffer should be empty";
+  ASSERT_FALSE(buffer.get_latest(obj1));
+}
+
+TEST(LockFreeMPMCQueue, test_bounded_push)
+{
+  {
+    LockFreeMPMCQueue<double, 10> queue;
+    ASSERT_TRUE(queue.empty()) << "Buffer should be empty";
+    for (auto i = 1; i <= 25; i++) {
+      ASSERT_TRUE(queue.bounded_push(i)) << "Buffer should have space for element as size is 10";
+      ASSERT_EQ(10, queue.capacity());
+    }
+    ASSERT_EQ(10, queue.capacity());
+
+    // when we start popping, the pop-ed elements should start from 16
+    for (auto i = 1u; i <= queue.capacity(); i++) {
+      double obj1;
+      ASSERT_TRUE(queue.pop(obj1));
+      ASSERT_EQ(i + 15, obj1);
+    }
+    double obj1;
+    ASSERT_FALSE(queue.pop(obj1));
+    ASSERT_TRUE(queue.empty()) << "Buffer should be empty";
+  }
+  {
+    LockFreeMPMCQueue<double> queue(10);
+    ASSERT_TRUE(queue.empty()) << "Buffer should be empty";
+    for (auto i = 1; i <= 25; i++) {
+      ASSERT_TRUE(queue.bounded_push(i)) << "Buffer should have space for element as size is 10";
+      ASSERT_EQ(10, queue.capacity());
+    }
+    ASSERT_EQ(10, queue.capacity());
+
+    // when we start popping, the pop-ed elements should start from 16
+    for (auto i = 1u; i <= queue.capacity(); i++) {
+      double obj1;
+      ASSERT_TRUE(queue.pop(obj1));
+      ASSERT_EQ(i + 15, obj1);
+    }
+    double obj1;
+    ASSERT_FALSE(queue.pop(obj1));
+    ASSERT_TRUE(queue.empty()) << "Buffer should be empty";
+  }
+}
+
+TEST(LockFreeMPMCQueue, test_lockfree_queue_push)
+{
+  LockFreeMPMCQueue<int, 100> queue;
+  int producer_count = 0;
+  std::atomic_int consumer_count(0);
+  std::atomic_bool done(false);
+
+  const int iterations = 1000000;
+
+  std::thread producer([&]() {
+    for (int i = 0; i < iterations; ++i) {
+      while (!queue.push(i)) {
+        std::this_thread::yield();
+      }
+      ++producer_count;
+    }
+  });
+
+  std::thread consumer([&]() {
+    int value;
+    while (!done) {
+      while (queue.pop(value)) {
+        ++consumer_count;
+      }
+    }
+    while (queue.pop(value)) {
+      ++consumer_count;
+    }
+  });
+
+  producer.join();
+  done = true;
+  consumer.join();
+
+  ASSERT_EQ(producer_count, consumer_count);
+  ASSERT_EQ(producer_count, iterations);
+  ASSERT_EQ(consumer_count, iterations);
+}
+
+TEST(LockFreeMPMCQueue, test_lockfree_queue_bounded_push)
+{
+  LockFreeMPMCQueue<int, 100> queue;
+  std::atomic_int producer_count = 0;
+  std::atomic_int consumer_count(0);
+  std::atomic_bool done(false);
+
+  const int iterations = 1000000;
+
+  std::thread producer([&]() {
+    for (auto j = 0; j < iterations; ++j) {
+      ASSERT_TRUE(queue.bounded_push(j));
+      ASSERT_EQ(100, queue.capacity());
+      ++producer_count;
+    }
+  });
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+  std::cerr << "producer_count: " << producer_count << std::endl;
+
+  std::thread consumer([&]() {
+    int value;
+    while (!done) {
```

Comment:
Can we merge those two conditions?
```suggestion
    while (!done && queue.pop(value)) {
```